File: ios/Cloudmoji/Cloudmoji/Views/WordBubble.swift
import SwiftUI

/// The `wordFloat` keyframes from `src/index.css`, as the four states the
/// animation actually interpolates between.
///
/// Written as a type rather than inline numbers so the curve can be asserted:
/// a bubble whose `entering` opacity is 1 never appears to arrive, and that is
/// invisible to any test that only checks the bubble draws.
enum WordBubblePhase: CaseIterable {
    /// `0% { opacity: 0; translateY(12px) scale(0.7) }`
    case entering
    /// `15% { opacity: 1; translateY(0) scale(1.06) }` — the overshoot.
    case arrived
    /// `25% { scale: 1 }` — settled, and held from here.
    case settled
    /// `100% { opacity: 0; translateY(-10px) scale(0.95) }`
    case leaving

    var opacity: Double {
        switch self {
        case .entering, .leaving: 0
        case .arrived, .settled: 1
        }
    }

    var scale: CGFloat {
        switch self {
        case .entering: 0.7
        case .arrived: 1.06
        case .settled: 1
        case .leaving: 0.95
        }
    }

    var offsetY: CGFloat {
        switch self {
        case .entering: 12
        case .arrived, .settled: 0
        case .leaving: -10
        }
    }
}

/// Every number the bubble is drawn and timed from.
///
/// The durations are the `wordFloat` percentages resolved against its 2.2s: the
/// percentages are kept as the source so a change to ``lifetime`` moves all four
/// legs together instead of leaving the fade-out hanging past the removal.
enum WordBubbleMetrics {
    /// `wordFloat 2.2s`, and the 2200ms timeout in
    /// `src/components/WordsMode.tsx` that removes the bubble. They are the same
    /// number on the web and must stay the same number here — a bubble that
    /// fades slower than its owner removes it disappears with a jump cut.
    static let lifetime: Double = 2.2

    static let arriveAt = 0.15 * lifetime   // 0.330
    static let settleAt = 0.25 * lifetime   // 0.550
    static let fadeFrom = 0.78 * lifetime   // 1.716

    static let enterDuration = arriveAt
    static let settleDuration = settleAt - arriveAt
    static let holdDuration = fadeFrom - settleAt
    static let exitDuration = lifetime - fadeFrom

    /// Design system type scale: word bubble is 18px / 900 / letter-spacing 0.5.
    static let fontSize: CGFloat = 18
    static let tracking: CGFloat = 0.5
    static let emojiSize: CGFloat = 22

    static let cornerRadius: CGFloat = 18
    static let horizontalPadding: CGFloat = 18
    static let verticalPadding: CGFloat = 5
    static let spacing: CGFloat = 6
    static let borderWidth: CGFloat = 1
}

/// The bubble itself, with no animation on it.
///
/// Split out because the animated wrapper starts at opacity 0 and only becomes
/// visible once `.task` has run — which `ImageRenderer` never does. Every pixel
/// assertion about colour and content therefore targets this, and the wrapper is
/// checked separately for actually becoming visible.
struct WordBubbleLabel: View {
    let emoji: String
    let word: String

    var body: some View {
        HStack(spacing: WordBubbleMetrics.spacing) {
            Text(emoji)
                .font(.system(size: WordBubbleMetrics.emojiSize))
            Text(word)
                .font(Theme.body(WordBubbleMetrics.fontSize, .black))
                .tracking(WordBubbleMetrics.tracking)
                .foregroundStyle(Theme.textPrimary)
        }
        .padding(.horizontal, WordBubbleMetrics.horizontalPadding)
        .padding(.vertical, WordBubbleMetrics.verticalPadding)
        .background(
            // `linear-gradient(135deg, rgba(255,107,107,0.2), rgba(78,205,196,0.2))`.
            // Coral into teal — the app's two brand colours meeting. A neutral
            // material here would read as a system alert rather than as the app
            // answering the child.
            LinearGradient(
                colors: [Theme.coral.opacity(0.2), Theme.teal.opacity(0.2)],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            ),
            in: shape
        )
        .overlay(shape.stroke(Color.white.opacity(0.1), lineWidth: WordBubbleMetrics.borderWidth))
    }

    private var shape: RoundedRectangle {
        RoundedRectangle(cornerRadius: WordBubbleMetrics.cornerRadius, style: .continuous)
    }
}

/// The floating label showing the word being spoken. Rises, holds, and fades
/// over ``WordBubbleMetrics/lifetime``.
///
/// The owner is expected to drop it after the same 2.2s — and to give it
/// `.id(...)` per word, the way the web passes `key={id}`, so a repeat of the
/// same word replays the animation from the top rather than sitting there.
/// `.task(id:)` covers the case where it does not.
struct WordBubble: View {
    let emoji: String
    let word: String

    @State private var phase: WordBubblePhase = .entering

    var body: some View {
        WordBubbleLabel(emoji: emoji, word: word)
            .opacity(phase.opacity)
            .scaleEffect(phase.scale)
            .offset(y: phase.offsetY)
            // It is a report on what just happened, not a control — the child
            // is already being told out loud.
            .allowsHitTesting(false)
            .accessibilityIdentifier("word-bubble")
            .task(id: "\(emoji)\u{1F}\(word)") {
                // Assigned unanimated, so a bubble reused for a new word starts
                // from the bottom again instead of easing out of `leaving`.
                phase = .entering
                withAnimation(.easeOut(duration: WordBubbleMetrics.enterDuration)) {
                    phase = .arrived
                }
                guard await sleep(WordBubbleMetrics.enterDuration) else { return }
                withAnimation(.easeInOut(duration: WordBubbleMetrics.settleDuration)) {
                    phase = .settled
                }
                guard await sleep(
                    WordBubbleMetrics.settleDuration + WordBubbleMetrics.holdDuration
                ) else { return }
                withAnimation(.easeInOut(duration: WordBubbleMetrics.exitDuration)) {
                    phase = .leaving
                }
            }
    }

    /// `false` if the wait was cancelled — the bubble was removed or the word
    /// changed, and the rest of the sequence must not run.
    private func sleep(_ seconds: Double) async -> Bool {
        do {
            try await Task.sleep(for: .seconds(seconds))
            return true
        } catch {
            return false
        }
    }
}

#Preview("Bubble") {
    ZStack {
        Theme.background.ignoresSafeArea()
        VStack(spacing: 24) {
            WordBubbleLabel(emoji: "🍎", word: "apple")
            WordBubbleLabel(emoji: "🐶", word: "いぬ")
            WordBubble(emoji: "🌈", word: "rainbow")
        }
    }
}

File: android/app/src/main/java/app/cloudmoji/android/ui/words/WordBubble.kt
package app.cloudmoji.android.ui.words

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.cloudmoji.android.ui.theme.CloudmojiBodyFont
import app.cloudmoji.android.ui.theme.Coral
import app.cloudmoji.android.ui.theme.Teal
import app.cloudmoji.android.ui.theme.TextPrimary
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Every number the bubble is drawn and timed from. Mirrors iOS
 * `WordBubble.swift`'s `WordBubbleMetrics` — the durations are the
 * `wordFloat` keyframe percentages resolved against [lifetimeMs], kept as the
 * source so the fade-out always finishes exactly as the owner removes the
 * bubble rather than being cut into a jump cut.
 */
object WordBubbleMetrics {
    /** `wordFloat 2.2s`, and `WordsViewModel.BUBBLE_HOLD_MS` — the same
     * number on both sides of the show/hide boundary. */
    const val lifetimeMs = 2200
    val arriveAtMs = (0.15 * lifetimeMs).toInt() // 330
    val settleAtMs = (0.25 * lifetimeMs).toInt() // 550
    val fadeFromMs = (0.78 * lifetimeMs).toInt() // 1716

    val fontSize = 18.sp
    val emojiSize = 22.sp
    val cornerRadius = 18.dp
    val horizontalPadding = 18.dp
    val verticalPadding = 5.dp
    val spacing = 6.dp
    val borderWidth = 1.dp
}

/**
 * The floating label showing the word being spoken. Rises, holds and fades
 * over [WordBubbleMetrics.lifetimeMs] — the owner ([app.cloudmoji.android.model.WordsViewModel])
 * is expected to drop it after the same span. Ported from iOS `WordBubble.swift`.
 *
 * [id] is the typed item's own identity, not its content: a repeat of the
 * exact same word must replay the float from the bottom rather than sit
 * there, which content alone cannot distinguish (the same emoji tapped twice
 * in a row is the same `(emoji, word)` pair).
 *
 * Decoration, never a control — it carries no `clickable`, so a tap aimed at
 * whatever is underneath (the emoji grid in landscape, where this floats as
 * an overlay) passes straight through, the same as iOS's `allowsHitTesting(false)`.
 */
@Composable
fun WordBubble(emoji: String, word: String, id: Long, modifier: Modifier = Modifier) {
    val alpha = remember { Animatable(0f) }
    val scale = remember { Animatable(0.7f) }
    val offsetY = remember { Animatable(12f) }

    LaunchedEffect(id) {
        alpha.snapTo(0f)
        scale.snapTo(0.7f)
        offsetY.snapTo(12f)

        // 0 -> arriveAt: opacity 0 -> 1, scale 0.7 -> 1.06 (the overshoot), offsetY 12 -> 0.
        launch { alpha.animateTo(1f, tween(WordBubbleMetrics.arriveAtMs, easing = LinearOutSlowInEasing)) }
        launch { scale.animateTo(1.06f, tween(WordBubbleMetrics.arriveAtMs)) }
        offsetY.animateTo(0f, tween(WordBubbleMetrics.arriveAtMs))

        // arriveAt -> settleAt: scale 1.06 -> 1 (settled, held from here).
        scale.animateTo(1f, tween(WordBubbleMetrics.settleAtMs - WordBubbleMetrics.arriveAtMs))

        // Held until fadeFrom.
        delay((WordBubbleMetrics.fadeFromMs - WordBubbleMetrics.settleAtMs).toLong())

        // fadeFrom -> lifetime: opacity 1 -> 0, offsetY 0 -> -10, scale 1 -> 0.95.
        launch { alpha.animateTo(0f, tween(WordBubbleMetrics.lifetimeMs - WordBubbleMetrics.fadeFromMs)) }
        launch { scale.animateTo(0.95f, tween(WordBubbleMetrics.lifetimeMs - WordBubbleMetrics.fadeFromMs)) }
        offsetY.animateTo(-10f, tween(WordBubbleMetrics.lifetimeMs - WordBubbleMetrics.fadeFromMs))
    }

    val shape = RoundedCornerShape(WordBubbleMetrics.cornerRadius)

    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(WordBubbleMetrics.spacing),
        modifier = modifier
            .graphicsLayer {
                this.alpha = alpha.value
                scaleX = scale.value
                scaleY = scale.value
                translationY = offsetY.value.dp.toPx()
            }
            .clip(shape)
            .background(
                // Coral into teal — the app's two brand colours meeting. A
                // neutral material here would read as a system alert rather
                // than as the app answering the child.
                Brush.linearGradient(listOf(Coral.copy(alpha = 0.2f), Teal.copy(alpha = 0.2f))),
            )
            .border(WordBubbleMetrics.borderWidth, Color.White.copy(alpha = 0.1f), shape)
            .padding(horizontal = WordBubbleMetrics.horizontalPadding, vertical = WordBubbleMetrics.verticalPadding)
            // It is a report on what just happened, not something to
            // navigate by — the child is already being told out loud.
            .clearAndSetSemantics {}
            .testTag("word-bubble"),
    ) {
        Text(text = emoji, fontSize = WordBubbleMetrics.emojiSize)
        Text(
            text = word,
            color = TextPrimary,
            fontFamily = CloudmojiBodyFont,
            fontWeight = FontWeight.Black,
            fontSize = WordBubbleMetrics.fontSize,
            letterSpacing = 0.5.sp,
        )
    }
}
